src/libopensc/card.c

```diff
@@ -412,9 +412,23 @@ int sc_disconnect_card(sc_card_t *card)
 
 	ctx = card->ctx;
 	LOG_FUNC_CALLED(ctx);
-
+	
 	if (card->lock_count != 0)
-		return SC_ERROR_NOT_ALLOWED;
+		LOG_FUNC_RETURN(ctx, SC_ERROR_NOT_ALLOWED);
+
+	sc_force_disconnect_card(card);
+	LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+}
+
+int sc_force_disconnect_card(sc_card_t *card)
```

Comment:
I'm not sure if handling this case globally is the right approach. Your commit message states the following:
>When performing context specific login with PIV card,
PIV driver does one extra sc_lock() call and expects
sc_unlock() to be called by following cryptographic operation.
If an error occurs in between, sc_disconnect_card will be called,
but will not release the card due to remaining locks.

So if I understand correctly, the additional lock is set and unset in different call backs of the PIV driver where `priv->context_specific` is handled. In this case, the card driver would be responsible for releasing resources that it has allocated. At the very least, `piv_finish()` could detect and release this additional lock, right @dengert ?

Reply:
It could work if  https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-piv.c#L3502-L3507 was copied 
after https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-piv.c#L3008

But `sc_card_disconnect`  needs these lines switched: 
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card.c#L416-L417
need to be after these lines:
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card.c#L416-L417

In the general case, if `sc_card_disconnect` is called with `card->lock_count != 0` should it just release the locks?

Reply:
Thank you for looking into this. As suggested, I removed the lock count check and added the unlock into `piv_finish`

Reply:
thank you